sklearn/neighbors/classification.py

```diff
@@ -310,13 +380,42 @@ def predict(self, X):
                                      'dataset')
 
         weights = _get_weights(neigh_dist, self.weights)
-
         if weights is None:
-            mode = np.asarray([stats.mode(pl)[0] for pl in pred_labels],
-                              dtype=np.int)
-        else:
-            mode = np.asarray([weighted_mode(pl, w)[0]
-                               for (pl, w) in zip(pred_labels, weights)],
-                              dtype=np.int)
+            # `neigh_dist` is an array of objects, where each
+            # object is a 1D array of indices.
+            weights = np.array([np.ones(len(row)) for row in neigh_dist])
 
-        return mode.flatten().astype(np.int)
+        probabilities = np.zeros((X.shape[0], self._classes.size))
+
+        # We cannot vectorize the following because of the way Python handles
+        # M += 1: if a predicted index was to occur more than once (for a
+        # given tested point), the corresponding element in `probabilities` 
+        # would still be incremented only once.
+        for i, pi in enumerate(pred_labels):
+            if len(pi) < 1:
+                probabilities[i] = 1e-6  # prevent division by zero later
+                continue  # outlier
+            # When we support NumPy >= 1.6, we'll be able to simply use:
+            # np.bincount(pi, weights, minlength=self._classes.size)
+            unpadded_probs = np.bincount(pi, weights[i])
+            probabilities[i] = np.append(unpadded_probs,
+                                         np.zeros(self._classes.size -
+                                                  unpadded_probs.shape[0]))
+
+        # Compute the unnormalized posterior probability, taking
+        # self.class_prior_ into consideration.
+        class_count = np.bincount(self._y)
```

Comment:
nowadays we need to handle multioutput y. Perhaps each will have its own prior...